grr/server/grr_response_server/gui/api_plugins/metadata_test.py

```diff
@@ -0,0 +1,382 @@
+    import openapi_spec_validator
+    try:
+      errors = openapi_spec_validator.validate_spec(self.openapi_desc_dict)
+    except:
+      errors = True
+
+    self.assertIsNone(errors)
+
+  def testAllRoutesInOpenApiDescription(self):
+    for router_method in self.router_methods.values():
+      for http_method, path, strip_root_types in router_method.http_methods:
+        simple_path = self.handler._SimplifyPath(path)
+        http_method = http_method.lower()
+
+        operation_obj = \
+          self.openapi_desc_dict["paths"][simple_path][http_method]
+
+        self.assertIsNotNone(operation_obj)
+
+  def testRouteArgsSeparationInPathQueryBody(self):
+    for router_method in self.router_methods.values():
+      for http_method, path, strip_root_types in router_method.http_methods:
+        simple_path = self.handler._SimplifyPath(path)
+        http_method = http_method.lower()
+
+        field_descriptors = []
+        if router_method.args_type:
+          field_descriptors = router_method.args_type.protobuf.DESCRIPTOR.fields
+
+        path_args = self.handler._GetPathArgsFromPath(path)
+        path_args = set(path_args)
+
+        # Triage parameters from the RDFProtoStruct class.
+        path_params = set()
+        query_params = set()
+        body_params = set()
+
+        for field_d in field_descriptors:
+          field_name = field_d.name
+          if field_name in path_args:
+            path_params.add(field_name)
+          elif http_method.upper() in ("GET", "HEAD"):
+            query_params.add(field_name)
+          else:
+            body_params.add(field_name)
+
+        # Triage parameters from the generated OpenAPI description.
+        openapi_path_params = set()
+        openapi_query_params = set()
+        openapi_body_params = set()
+
+        # https://swagger.io/specification/#operation-object
+        operation_obj = \
+          self.openapi_desc_dict["paths"][simple_path][http_method]
+
+        # parameters = operation_obj.get("parameters")
+        # self.assertIsNotNone(parameters)
+        parameters = operation_obj["parameters"]
+        for parameter in parameters:
+          param_name = parameter["name"]
+          if parameter["in"] == "path":
+            openapi_path_params.add(param_name)
+          elif parameter["in"] == "query":
+            openapi_query_params.add(param_name)
+          else:
+            raise TypeError("Wrong OpenAPI parameter location: "
+                            "\"{parameter["in"]}\"")
+
+        request_body_obj = operation_obj.get("requestBody")
+        if request_body_obj:
+          schema_obj = request_body_obj["content"]["application/json"]["schema"]
+          for param_name in schema_obj["properties"]:
+            openapi_body_params.add(param_name)
+
+        self.assertSetEqual(path_params, openapi_path_params)
+        self.assertSetEqual(query_params, openapi_query_params)
+        self.assertSetEqual(body_params, openapi_body_params)
```

Comment:
I think this test might be improved. The main problem: the test contains loops and a lot of conditional logic. This means that it's somewhat hard to read, but also that it might have hard to spot bugs.

What this test seems to be doing is checking that openapi spec path, query and body params are figured out correctly. Instead of using a generic testing logic, I'd refactor this test to be as simple as possible: something like a series of assertEqual calls (note that assertEqual works fine on complex nested data structures). Here the dummy router is used, meaning that we know the API methods we work with and the expected results for each method.

Reply:
Thank you for pointing this out! I made the change to explicitly written literals throughout the testing code. Please let me know if there's something that needs to be changed.

Reply:
@alexandrucosminmihai - thanks for the latest changes! I did another pass at reviewing the testing code. Will resolve this comment, so that we can continue the conversation in the latest round of comments.